### ISSUE TITLE
[Bug/Question] Shell routing/navigation reuses the same page instance on every call to Shell.GoToAsync()

### ISSUE THREAD

**Issue by chrisstaley** [xamarin/Xamarin.Forms#7555]
### Description
The Shell's navigation logic appears to only ever create a single instance of the `Page` registered to a given route, and then reuses it for all navigation requests to that route.

I'm not sure if this is a bug, or if it's intentional and I'm just using it incorrectly. I imagine it's the latter, so further direction would be appreciated.

### Steps to Reproduce
Given the attached sample project:
[ShellNavRepro.zip](https://github.com/xamarin/Xamarin.Forms/files/3623066/ShellNavRepro.zip)

1. Run the app and observe that `MainPage` is initialized and displayed.
2. Tap the "Another" link in the flyout menu, and observe that the `AnotherPage` is initialized and displayed.
3. Tap the "Main" link in the flyout menu, and observe that the same instance of `MainPage`created in step 1 is shown.
4. Tap the "Another" link in the flyout menu, and observe that the same instance of `AnotherPage` created in step 2 is shown.

### Expected Behavior
I expected new instances of pages to be created any time their menu items are tapped (i.e. their routes are navigated to). Likewise, whenver a screen is navigated away from, I expected the page to be discarded/disposed.

There are at least a couple of issues that I'm having with this behavior:
1. I can't tell the difference between when a screen is displayed for the first time, and when it is being displayed again after foregrounding the app (the `Page.OnAppearing()` method is called either way).
2. Pages with route parameters (i.e. via the `QueryPropertyAttribute`) can end up in states where you don't know if the parameters were set from the current route navigation attempt or a previous attempt, since route parameters are only set/reset/cleared if they are specified. So invoking a route with *no parameters* after having previously invoked the route *with parameters* leaves the parameters set (i.e. since the same `Page` instance is used), 

### Actual Behavior
Page instances are reused between separate navigation attempts.

### Basic Information
- Version with issue: 4.2.0.778463

**Comment by PureWeen** [comment]
@chrisstaley 

This is all by design. 
We don't want to start a page from the inception point when you switch between flyouts items. The renderers attached to the page do get disposed but the cross platform pages themselves retain the state so users can return to where they are left off

> I can't tell the difference between when a screen is displayed for the first time, and when it is being displayed again after foregrounding the app (the Page.OnAppearing() method is called either way).

We'd like to create some better life cycle events at some point :-/ we just haven't quite perfected the API on this one. Things like Loaded or Rendered etc...   At this point I'd recommend using a bool with OnAppearing to recognize the first time it shows up

> Pages with route parameters (i.e. via the QueryPropertyAttribute) can end up in states where you don't know if the parameters were set from the current route navigation attempt or a previous attempt, since route parameters are only set/reset/cleared if they are specified. So invoking a route with no parameters after having previously invoked the route with parameters leaves the parameters set (i.e. since the same Page instance is used),

"So invoking a route with no parameters after having previously invoked the route with parameters leaves the parameters set (i.e. since the same Page instance is used),"

Does it? I could have sworn we fixed this so that it just sets it to null if you don't specify a query parameter. Can you send me a reproduction of this scenario?

**Comment by chrisstaley** [comment]
> We don't want to start a page from the inception point when you switch between flyouts items.

It would be nice to have a hook to control that, since, depending on the app, this behavior may or may not make sense. As a completely contrived example, given a survey screen wherein navigating away meant abandoning the survey, coming back to the screen should mean restarting the survey. Having a way to handle this natively would prevent the need for roll-your-own code from being copied around between projects.

This could also perhaps be implemented as a place where dependency resolution could be hooked in, which is something I've seen requested more than once.

> We'd like to create some better life cycle events at some point

+1

>  Can you send me a reproduction of this scenario?

[Issue7555.zip](https://github.com/xamarin/Xamarin.Forms/files/3697886/Issue7555.zip)

Repro steps:
1. Start app.
2. Click "Another (MenuItem with param)" link in flyout menu. (Observe: param value is reported as '1', which is expected.)
3. Click "Main" link in flyout menu.
4. Click "Another (FlyoutItem without param)" link in flyout menu.

Results:
	1. Param value is still reported as '1', even though no parameter was specified (albeit implicitly since FlyoutItem link was used which does not have a parameter).
	2. After Step 2, the wrong menu item was highlighted in the flyout menu (should submit another issue for this?).

**Comment by jpveldtman** [comment]
This seems related to my issue #8347, please confirm?

**Comment by hartez** [comment]
> This seems related to my issue #8347, please confirm?

They don't appear to be the same issue.

**Comment by duncancole** [comment]
I'm also having trouble with this issue - I have various pages that are reused for similar content, for example a page that display a video and some text. With this current implementation if  you got a video page, back out and select a different video it takes you to that same page and the video you previously selected starts playing again until the new video is loaded (there is a delay of up to a couple of seconds) which is not ideal. I also wanted to stop the 'load video' logic that I have triggering OnAppearing if the video is already loaded so if the user goes fullscreen and then closes that the video doesn't, well, load again. However because the screen is completely reused when you go back to a video I can't be sure of what context OnAppearing is happening.

This feels like a pretty standard use case for any app that displays content so I'm wondering if I'm missing something here?

**Comment by brunck** [comment]
@duncancole I tend to have a "third state" for a page like this where it shows nothing, or a loading spinner, or something like that, until the actual content is loaded. Not having such a thing caused problems with UI testing and I found it was useful for more than just that. I think there is something in the Xamarin Community Toolkit around this scenario.
